### ISSUE TITLE
PJSUA2: inconsistence and missing calls of method onCallTsxState

### ISSUE THREAD

**Issue by wosrediinanatour** [pjsip/pjproject#3605]
### Describe the bug

Documentation says the following
 ```c
  1 typedef enum pjsip_tsx_state_e
  2 {
  3     PJSIP_TSX_STATE_NULL,       /**< For UAC, before any message is sent.   */
  4     PJSIP_TSX_STATE_CALLING,    /**< For UAC, just after request is sent.   */
  5     PJSIP_TSX_STATE_TRYING,     /**< For UAS, just after request is received.*/
  6     PJSIP_TSX_STATE_PROCEEDING, /**< For UAS/UAC, after provisional response.*/
  7     PJSIP_TSX_STATE_COMPLETED,  /**< For UAS/UAC, after final response.     */
  8     PJSIP_TSX_STATE_CONFIRMED,  /**< For UAS, after ACK is received.        */
  9     PJSIP_TSX_STATE_TERMINATED, /**< For UAS/UAC, before it's destroyed.    */
 10     PJSIP_TSX_STATE_DESTROYED,  /**< For UAS/UAC, will be destroyed now.    */
 11     PJSIP_TSX_STATE_MAX         /**< Number of states.                      */
 12 } pjsip_tsx_state_e;
```

For incoming SIP INVITEs (initial and reINVITEs):
* onCallTsxState with PJSIP_TSX_STATE_TRYING is not called
* (whereas for outgoing initial SIP INVITES, onCallTsxState  with PJSIP_TSX_STATE_CALLING is called)

In case of outgoing hang_up:
 * onCallTsxState is never called
 
For outgoing SIP INVITEs (UAC) that are CANCEL'ed:
 * onCallTsxState not called for CANCEL
 * There isn't any onCallTsxState with PJSIP_TSX_STATE_COMPLETED or PJSIP_TSX_STATE_TERMINATED

For incoming SIP BYE:
 * onCallTsxState is called twice with PJSIP_TSX_STATE_COMPLETED, the first carries the response, the second carries the request!

Other SIP transaction seems to be ok (e.g. SIP INFO):
 * PJSIP_TSX_STATE_CALLING 
 * PJSIP_TSX_STATE_TRYING
 * PJSIP_TSX_STATE_COMPLETED

### Steps to reproduce

Implement onCallTsxState(...) and log when it is called.

### PJSIP version

2.13

### Context

Linux

### Log, call stack, etc

```shell
(Invokation of onCallTsxState (...) is never logged)
```

**Comment by wosrediinanatour** [comment]
(updated description with behavior of CANCEL)

**Comment by sauwming** [comment]
For hangup/cancel:
This is due to the immediate disconnection feature (the last bullet point in https://github.com/pjsip/pjproject/issues/1049). So call is declared to be disconnected immediately and any PJSUA call related callback will no longer be called).

For incoming INVITE:
You're right. This is a problem described in #530. Ideally we should fix this, but I'm not sure whether it causes any issue though.

For incoming BYE:
I tested it here with the latest master and it seems to be okay (tsx state called only once with state COMPLETED). Perhaps you can attach the log here so we can check?

**Comment by wosrediinanatour** [comment]
Thank you for your answer!

For hangup/cancel: It is somehow inconsistent that there is no `onCallTsxState(...)` called, but `onCallState(...)`. So any app that uses `onCallTsxState(...)` has also to "observe" `onCallState(...)`.

For incoming INVITE: Which proble could occur? Since `onCallState(...)` is called with a "trying" state, here it would be sane/logical/assumed.

For incoming BYE: Having two calls, one with the incoming BYE message, and one with the outgoing 200 OK message is actually quite nice. But: The order is somehow confusing and may make apps more complicated.

**Comment by sauwming** [comment]
For outgoing hangup/cancel: Since app is the one initiating the hangup, observing `onCallState()` is actually not  necessary. Once you invoke the hangup API, call can be considered to have been disconnected.

There are indeed pros and cons of the immediate disconnection feature. The con, is as you mentioned, you will lose some callbacks. But the pro is that you can immediately cleanup your resources without waiting for all the callbacks to be called, which can take a long time, especially if remote is non-responsive. The callbacks after call is disconnected are usually not useful anyway.

For incoming BYE: As previously mentioned, my local testing only gave me one callback, so I'm not sure why you get two callbacks. If you can attach your log, I can probably check the reason.

For incoming INVITE: I tried to fix this, but it turns out to be rather complicated. The problem is that when we receive the INVITE request, `onIncomingCall()` hasn't been called and the call hasn't been created, so we can't pass the `PJSIP_TSX_STATE_TRYING` event to the call. The tsx state itself is there, if you use log level 5, you will see:
```
17:51:26.841         tsx0x10a591128  ..Incoming Request msg INVITE/cseq=16836 (rdata0x109f13808) in state Null
17:51:26.841         tsx0x10a591128  ...State changed from Null to Trying, event=RX_MSG
```
it's just the Call object hasn't existed.

In summary, my opinion is as follows:
- The documentation of `pjsip_tsx_state_e` is correct. It is for the PJSIP level, i.e. the SIP transaction itself. It will always undergo the correct state transitions. You can use PJSIP log level 5 to check the details.
- For `onCallTsxState()`, the doc says: `This is a general notification callback which is called whenever a transaction within the call has changed state.` This is also not wrong, IMO, the keyword here is `transaction within the call`. So if the call hasn't been created (in the case of incoming INVITE) or has already been destroyed (in the case of hangup), you naturally won't get the callbacks. In other words, the callback will notify you when a transaction changes state during the lifetime of the Call object.

**Comment by wosrediinanatour** [comment]
Your explanation is sound...
but I am not very happy from a user perspective. :-)

Btw: When receiving a BY in 2.13 (not master) I get (note that I had to add the log):
```
State changed from CONFIRMED to DISCONNCTD, event=TSX_STATE
State changed from CONFIRMED to DISCONNCTD, event=TSX_STATE
``` 
Both lines are shown after `onCallTsxState()`, the second even after `onCallStateChanged()`.
The first `onCallTsxState()` has a tx_message message, the second a rx_message.

**Comment by wosrediinanatour** [comment]
Ad SIP ReINVITE - there `onCallTsxState(...)` with PJSIP_TSX_STATE_TRYING  is *not* called either:
```
Received Request msg INVITE/cseq=4096 (rdata0x62500006e128)
Transaction created for Request msg INVITE/cseq=4096 (rdata0x62500006e128)
Incoming Request msg INVITE/cseq=4096 (rdata0x62500006e128) in state Null
State changed from Null to Trying, event=RX_MSG
 Transaction tsx0x61d000047f28 state changed to Trying
Got SDP offer in Request msg INVITE/cseq=4096 (rdata0x62500006e128)
Call 2: received updated media offer 
... onCallRxReinvite()...
Response msg 100/INVITE/cseq=4096 (tdta0x6210001401a8) created
Initial answer Response msg 100/INVITE/cseq=4096 (tdta0x6210001401a8)
Sending Response msg 100/INVITE/cseq=4096 (tdta0x6210001401a8)
Sending Response msg 100/INVITE/cseq=4096 (tdta0x6210001401a8)
Sending Response msg 100/INVITE/cseq=4096 (tdta0x6210001401a8) in state Trying
... 100 Trying sent ....
State changed from Trying to Proceeding, event=TX_MSG
Transaction tsx0x61d000047f28 state changed to Proceeding
... onCallTsxState() called with TX_MSG ...
```

**Comment by sauwming** [comment]
I have created PR #3614 which should fix the missing `PJSIP_TSX_STATE_TRYING` event for re-INVITE and the duplicate `PJSIP_TSX_STATE_COMPLETED` events for BYE.

Please check it out and let us know if you have any feedbacks.
